crates/ra_hir_ty/src/tests.rs

```diff
@@ -33,7 +33,9 @@ fn type_at_pos(db: &TestDB, pos: FilePosition) -> String {
     let expr = algo::find_node_at_offset::<ast::Expr>(file.syntax(), pos.offset).unwrap();
     let fn_def = expr.syntax().ancestors().find_map(ast::FnDef::cast).unwrap();
     let module = db.module_for_file(pos.file_id);
-    let func = module.child_from_source(db, InFile::new(pos.file_id.into(), fn_def)).unwrap();
+    let func = *module.child_by_source(db)[keys::FUNCTION]
+        .get(&InFile::new(pos.file_id.into(), fn_def))
```

Comment:
I don't quite understand why the API is always building a whole map when it seems like all uses just immediately do `get` on it (maybe I missed something)

Reply:
Building the whole map is easier: you just traverse everything and put it into the map. If you were directly looking for a specific Item, you would have first filter by item type and, in general, provide separate code for each type of looked for item. See the deleted `child_from_source` module, it pursued that approach. 

The map should also be cache able in theory, so that repeated lookups for neighboring nodes will be fast.

Reply:
Maybe it makes sense to change `() -> DynMap` to `(&mut DynMap)` though: that way, we can change the API in the future to use a `DynMap` which drops unrelated keys on the floor.